flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java

```diff
@@ -101,20 +114,26 @@ public DataInputStatus calculateOverallStatus() throws IOException {
     }
 
     void nextSelection() {
-        if (inputSelectable == null) {
-            inputSelection = InputSelection.ALL;
+        if (inputSelectable == null || areAllDataInputsFinished()) {
+            selectedInputsMask = InputSelection.ALL.getInputMask();
+        } else if (dataFinishedButNotPartition != 0) {
+            selectedInputsMask =
+                    (inputSelectable.nextSelection().getInputMask() | dataFinishedButNotPartition)
+                            & allSelectedMask;
         } else {
-            inputSelection = inputSelectable.nextSelection();
+            selectedInputsMask = inputSelectable.nextSelection().getInputMask();
```

Comment:
This is adding quite a bit extra checks on the hot path. I have a feeling we could streamline it by adding some explicit state transition enum, especially on the `inputSelectable != null` path. `NORMAL` -> `NOT_ALL_DATA_INPUTS_FINISHED` -> `ALL_DATA_INPUTS_FINISHED`?. We would still have a switch case instead of single if, but it should keep single check for `inputSelectable == null` hot path, and just add a single extra check for `NORMAL` 
(pseudo code)
```
switch (...) { 
  case inputSelectable==null:
    foo();
  case NORMAL:
    bar();
  (...)
}
```
I would guess those are two simple if checks, while currently for `NORMAL` you have to evaluate the following condition:
```
!(inputSelectable == null || areAllDataInputsFinished()) && dataFinishedButNotPartition == 0
```

Reply:
I took a slightly different approach and I recalculate the `areAllDataInputsFinished` only when I receive the `END_OF_DATA`. It should be effectively very similar to what you're suggesting.

Reply:
Why don't you go one step further and squash `allDataInputsFinished` with `dataFinishedButNotPartition != 0`  and `inputSelectable == null` into a single enum check? Something like 
```
private OperatingMode mode = PROCESSING_NO_INPUT_SELECTABLE | PROCESSING | NOT_ALL_DATA_INPUTS_FINISHED | ALL_DATA_INPUTS_FINISHED
```
? It would be one comparison faster and I think cleaner?

Reply:
Per my personal taste it's not cleaner, but I am not arguing against faster. I updated it as per your suggestion.